DifferentiationInterface/ext/DifferentiationInterfaceMooncakeExt/differentiate_with.jl

```diff
@@ -43,40 +54,50 @@ end
 function Mooncake.rrule!!(dw::CoDual{<:DI.DifferentiateWith}, x::CoDual{<:AbstractArray})
     primal_func = primal(dw)
     primal_x = primal(x)
-    fdata_arg = fdata(x.dx)
+    fdata_arg = x.dx
     (; f, backend) = primal_func
     y = zero_fcodual(f(primal_x))
 
     # output is a vector, so we need to use the vector pullback
     function pullback_array!!(dy::NoRData)
-        tx = DI.pullback(f, backend, primal_x, (fdata(y.dx),))
-        @assert first(only(tx)) isa rdata_type(typeof(first(primal_x)))
+        tx = DI.pullback(f, backend, primal_x, (y.dx,))
+        @assert rdata(first(only(tx))) isa rdata_type(tangent_type(typeof(first(primal_x))))
         fdata_arg .+= only(tx)
         return NoRData(), dy
     end
 
     # output is a scalar, so we can use the scalar pullback
     function pullback_scalar!!(dy::Number)
         tx = DI.pullback(f, backend, primal_x, (dy,))
-        @assert first(only(tx)) isa rdata_type(typeof(first(primal_x)))
+        @assert rdata(first(only(tx))) isa rdata_type(tangent_type(typeof(first(primal_x))))
         fdata_arg .+= only(tx)
         return NoRData(), NoRData()
     end
 
     # output is a Tuple, NTuple
     function pullback_tuple!!(dy::Tuple)
         tx = DI.pullback(f, backend, primal_x, (dy,))
-        @assert first(only(tx)) isa rdata_type(typeof(first(primal_x)))
+        @assert rdata(first(only(tx))) isa rdata_type(tangent_type(typeof(first(primal_x))))
         fdata_arg .+= only(tx)
         return NoRData(), NoRData()
     end
 
-    pullback = if typeof(primal(y)) <: Number
+    # inputs are non Differentiable
+    function pullback_nodiff!!(dy::NoRData)
+        @assert tangent_type(typeof(primal(x))) <: Vector{NoTangent}
+        return NoRData(), dy
+    end
+
+    pullback = if tangent_type(typeof(primal(x))) <: Vector{NoTangent}
+        pullback_nodiff!!
+    elseif typeof(primal(y)) <: Number
         pullback_scalar!!
-    elseif typeof(primal(y)) <: Array
+    elseif typeof(primal(y)) <: AbstractArray
         pullback_array!!
-    else
+    elseif typeof(primal(y)) <: Tuple
         pullback_tuple!!
+    else
+        error("$(typeof(primal(y))) primal type currently not supported.")
```

Comment:
Similarly, can you also print the type information of f and x, which are helpful for debugging?